external/llvm-project/mlir/lib/Dialect/AMDGPU/IR/AMDGPUDialect.cpp

```diff
@@ -326,6 +326,62 @@ LogicalResult MFMAOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// DPPOp
+//===----------------------------------------------------------------------===//
+LogicalResult DPPOp::verify() {
+  Type srcType = getSrc().getType();
+  if (srcType.getIntOrFloatBitWidth() > 32) {
+    return emitOpError("integer and floating point types larger than 32 bits "
+                       "are not supported");
+  }
+
+  DPPPerm kind = getKind();
+  auto permArgument = getPermArgument();
+
+  switch (kind) {
+
+  case DPPPerm::quad_perm:{
+    auto quadPermAttr = permArgument->dyn_cast<ArrayAttr>();
+    if (!quadPermAttr || quadPermAttr.size() != 4) {
+      return emitOpError("quad_perm attribute must have exactly 4 elements");
+    }
+    for (int i = 0; i < 4; ++i) {
+      auto intAttr = quadPermAttr[i].dyn_cast<IntegerAttr>();
+      int num = intAttr.getInt();
+      if (num < 0 || num > 3) {
+        return emitOpError(
+            "Each element of quad_perm must be in the range [0, 3]");
+      }
+    }
+  }
+    break;
+
+  case DPPPerm::row_shl:
+  case DPPPerm::row_shr:
+  case DPPPerm::row_ror:{
+    if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+      int attrValue = intAttr.getInt();
+      if (attrValue < 1 || attrValue > 15) {
+        return emitOpError("Attribute value must be between 1 and 15");
+      }
+    }
+  }
+  break;
+
+  case DPPPerm::wave_shl:
+  case DPPPerm::wave_shr:
+  case DPPPerm::wave_rol:
+  case DPPPerm::wave_ror:
+  case DPPPerm::row_mirror:
+  case DPPPerm::row_half_mirror:
+  case DPPPerm::row_bcast_15:
+  case DPPPerm::row_bcast_31:
+
```

Comment:
Let's defensively make sure no one's trying to pass non-trivial arguments to these (that is, if an attribute's present, it's a UnitAttr)